### PR TITLE
test(rules): rstest framework

### PR DESCRIPTION
**Pull request by supermaxiste**
## Proposed Changes

Replaced previous rule tests with new tests following the [`rstest`](https://lib.rs/crates/rstest) framework. Now we have 1 test function per rule, with each function testing different scenarios.
Example of previous code structure:
```rust
#[test]
    fn rule1_example1() {
    mask = rule1(example1)
    assert!(mask.is_set())
}
#[test]
    fn rule1_example2() {
    mask = rule1(example2)
    assert!(mask.is_set())
}
```

Example of ✨ new ✨ code structure:
```rust
#[rstest]
#[case(example1)]
#[case(example2)]
   fn rule1(case)  {
   mask = rule1(case)
   assert!(mask.is_set())
}
```

Addresses #25 

## Types of Changes

- [ ] A **bug fix** (non-breaking change which fixes an issue).
- [x] A new **feature** (non-breaking change which adds functionality).
- [ ] A **breaking change** (fix or feature that would cause existing
      functionality to not work as expected).
- [ ] A **non-productive** update (documentation, tooling, etc. if none of the
      other choices apply).

## Checklist

- [x] I have read the
      [CONTRIBUTING](https://github.com/sdsc-ordes/tripsu/blob/master/CONTRIBUTING.md)
      guidelines.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added the necessary documentation (if appropriate).

### CODE REVIEW

File: src/rules.rs [hunk 1]
```diff
@@ -146,193 +147,105 @@ mod tests {
...
+    // Subject not in type index
+    #[case("Alice", "hasName", "Bob", "hasAge", "Bob", "Person", false, false)]
+    fn subject_predicate_rule(
+        #[case] subject_iri: &str,
+        #[case] predicate_iri: &str,
+        #[case] rule_subject: &str,
+        #[case] rule_predicate: &str,
+        #[case] index_subject: &str,
+        #[case] index_object: &str,
+        #[case] expected_s: bool,
```
Comment (cmdoret) [inline src/rules.rs:230]: **suggestion:** when we test the sp rule, we never want to mask the subject (do we?). If so, we could simplify this by removing subject_iri from the arguments and the test cases, while still keeping the assert to check the subject mask.

File: src/rules.rs [hunk 2]
```diff
@@ -146,193 +147,105 @@ mod tests {
...
-
-    #[test]
-    fn predicate_not_in_config() {
+        assert_eq!(mask.is_set(&TripleMask::SUBJECT), expected_s);
+        assert_eq!(mask.is_set(&TripleMask::OBJECT), expected_o);
+    }
+
+    #[rstest]
+    // Predicate is in the rules
+    #[case("hasName", "hasName", false, true)]
```
Comment (cmdoret) [inline src/rules.rs:189]: **suggestion:** don't we expect `expected_s` to always be false here? if so, same suggestion as below

File: src/rules.rs [hunk 3]
```diff
@@ -146,193 +147,105 @@ mod tests {
...
+    #[rstest]
+    // Subject is in the rules & type index
+    #[case(true, "Alice", "Alice", "Person", "Person", true, false)]
+    // Subject is in the type index, not in the rules
+    #[case(true, "Alice", "Alice", "Person", "Bank", false, false)]
+    // Subject is not in the type index
+    #[case(true, "Alice", "BankName", "Bank", "Bank", false, false)]
+    // Object is in the rules & type index
+    #[case(false, "Alice", "Alice", "Person", "Person", false, true)]
+    fn type_rule(
```
Comment (cmdoret) [inline src/rules.rs:159]: **suggestion:** same as  below, if `expected_o` is never be affected by this rule, we could drop arguments.

**Review by cmdoret — COMMENTED** [review]
This is much better and a perfect fit for this use-case!

I proposed to remove constant arguments with the objective of improving readability. Let me know if you think this is not a good idea.

**Comment by supermaxiste** [comment]
Thank you @cmdoret and very good suggestion!

For 2 out of 3 rules I left a simple assert and removed the argument that were constant across cases 👍 
For the `type_rule` we do test both subjects and objects being either masked or not, so I left the cases as they are.

**Review by cmdoret — APPROVED** [review]
:raised_hands:

**Pull request merged**